### DISCUSSION TITLE
Question on extra type casting to enable compilation

### DISCUSSION THREAD

**Discussion by greenozon** [icsharpcode/ILSpy#3321] [General]
Quick question: is it possible to somehow tell ILSpy during decompilation that certain casting is needed, otherwise it won't compile?

eg:

ILSPy decompiled some routine as follows, but it does not compile
![image](https://github.com/user-attachments/assets/c0819cc6-7584-4fb4-ae31-3e5ec9df3e45)

and IL for it:
![image](https://github.com/user-attachments/assets/a659b1ac-8e2d-46e1-b88e-917bf2f5b295)

Original code looks like

![image](https://github.com/user-attachments/assets/f05259e1-a007-4f9b-b5f9-774205aae01a)

**Comment by dgrunwald** [comment]
There's currently no option to force extra casts; we try to insert them only where needed.

ILSpy has its own C# compiler logic (`CSharpResolver`) that performs overload resolution and checks if writing `a == b` hits the correct comparison operator (if not, we'll insert casts). Looks like there's a bug in this particular case where our operator resolution logic believes that writing `a == b` ends up using the base class `object == object` comparison; but the C# compiler refuses to use any comparison (maybe it detects this as some kind of ambiguity?). -> Bug #3322
